src/haz3lcore/dynamics/Elaborator.re

```diff
@@ -261,6 +262,48 @@ let rec dhexp_of_uexp =
         let* c_fn = dhexp_of_uexp(m, fn);
         let+ c_arg = dhexp_of_uexp(m, arg);
         DHExp.Ap(c_fn, c_arg);
+      | DeferredAp(fn, arg) =>
+        let* c_fn = dhexp_of_uexp(m, fn);
+        switch (err_status) {
+        | InHole(ErroneousPartialAp(Meaningless)) => Some(c_fn)
+        | InHole(ErroneousPartialAp(ArityMismatch(_))) =>
+          Some(DHExp.InvalidText(id, 0, "<inv partial ap>"))
+        | _ =>
+          let mk_tuple = (ctor, xs) =>
+            List.length(xs) == 1 ? List.hd(xs) : ctor(xs);
+          let args = Term.UExp.matched_args(arg);
+          let* ty_fn = fixed_exp_typ(m, fn);
+          let (ty_arg, ty_ret) = Typ.matched_arrow(ty_fn);
+          let ty_ins = Typ.matched_args(List.length(args), ty_arg);
+          /* Substitute all deferrals for new variables */
```

Comment:
> there's no actual data dependency between elements here (i think) so instead of folding, just map, and then use OptUtil.sequence (see other examples in this file)

I gave it a try. The issue I encountered is that lines 206-207 require both pats and d_args to share the same name, and this name is determined by the length of the growing pats, so perhaps using a fold would be more suitable.

Reply:
Makes sense